docs: Add upgrade guide for Meilisearch plugin (#3583)
medusajs/medusa: docs/content/upgrade-guides/plugins/meilisearch/1-0-0.md

# Meilisearch Search: v1.0.0

> Version 1.0.0 of our official Meilisearch plugin comes with breaking changes to the plugin options that are passed to the Meilisearch service through `medusa-config.js`.

Reviewer: ```suggestion
Version 1.0.0 of the official Meilisearch plugin comes with breaking changes to the plugin options that are passed to the Meilisearch service through `medusa-config.js`.
```

## Overview

In the new version of the Meilisearch search plugin, two new plugin configuration properties are introduced; `transformer` and `primaryKey`.  As a result, the way indexes in Meilisearch are configured has changed. 

The new object shape for the plugin options is as follows:

```jsx
{
  config: {
		host: string
		apiKey: string
	}
  settings: {
    [indexName]: {
      indexSettings: {
        searchableAttributes: string[]
        displayedAttributes: string[]
				...
      }
      transformer?: <T>(medusaEntity: T) => unknown
    }
  }
}

> ```

Reviewer: As we have ESLint configured on languages like `js`, `jsx`, `ts`, and `tsx`, this block produces errors related to JavaScript. I suggest that we change the block to the following:

~~~suggestion
```js title=medusa-config.js
const plugins = [
  // ...
  {
    config: {
      host,
      apiKey
    }
    settings: {
      indexName: {
        indexSettings: {
          searchableAttributes,
          displayedAttributes
          // ...
        }
        transformer?: (product) => ({
          id: product.id,
          // ...
        })
      }
    }
  }
]
```
~~~

And then we can link to the MeiliSearch plugin for more details on what each of these mean, although I don't think that's necessary considering you already cover it later in the upgrade guide.

## Actions required

> As you can see from the new object shape, the property `indexSettings` has been introduced to hold the settings specific to Meilisearch’s index options. This has been done to make space for the `transformer` and `primaryKey`. 

Reviewer: ```suggestion
As you can see from the new object shape, the property `indexSettings` has been introduced to hold the settings specific to Meilisearch’s index options. This has been done to make space for new settings like `transformer` and `primaryKey`. 
```

> Previously, you might have configured the Meilisearch plugin as seen below:

Reviewer: ```suggestion
Previously, your Meilisearch plugin configurations were something like this:
```

```jsx
{
  config: {
		host: "https://search-api-example.com",
    apiKey: "some_key",
	}
  settings: {
    products: {
      searchableAttributes: ["title", "description"],
      attributesToRetrieve: ["title", "description"],
		}
  }
}

> ```

Reviewer: ~~~suggestion
```js title=medusa-config.js
const plugins = [
  // ...
  {
    config: {
      host: "https://search-api-example.com",
      apiKey: "some_key",
    },
    settings: {
      products: {
        searchableAttributes: ["title", "description"],
        attributesToRetrieve: ["title", "description"],
      }
    }
  }
]
```
~~~

> In the above example, an index `products` has been configured with two options. Updating to 1.0.0 requires you to nest these within the `indexSettings`. 

Reviewer: ```suggestion
In the above example, an index `products` has been configured with the two options `searchableAttributes` and `attributesToRetrieve`. Updating to 1.0.0 requires you to nest these options within the `indexSettings`. 
```

The updated plugin options would look like so:

```jsx
{
  config: {
		host: "https://search-api-example.com",
    apiKey: "some_key",
	}
  settings: {
    products: {
			indexSettings: {
	      searchableAttributes: ["title", "description"],
	      attributesToRetrieve: ["title", "description"],	
			}
		}
  }
}

> ```

Reviewer: ~~~suggestion
```js
const plugins = [
  // ...
  {
    config: {
      host: "https://search-api-example.com",
      apiKey: "some_key",
    },
    settings: {
      products: {
        indexSettings: {
          searchableAttributes: ["title", "description"],
          attributesToRetrieve: ["title", "description"],	
        }
      }
    }
  }
]
```
~~~

> You can read more about the new plugin options properties in the package README.

Reviewer: ```suggestion
You can learn more about the new settings in the [MeiliSearch plugin documentation](../../../plugins/search/meilisearch.md)
```
